### PR TITLE
Adaptive thresholds are useful

### LINKED ISSUE #3571: SHOULD implement adaptive packet threshold loss detection
The RACK draft has this as a SHOULD, and this was worthwhile in our deployment experience, so I think we should change our MAY to a SHOULD.

From Section 4 of https://tools.ietf.org/html/draft-ietf-tcpm-rack-08
"3.  The RACK reordering window SHOULD leverage Duplicate Selective
       Acknowledgement (DSACK) information [RFC3708] to adaptively
       estimate the duration of reordering events."

### PR DESCRIPTION
**Pull request by ianswett**
Fixes #3571 by tweaking the text, but doesn't add a new SHOULD.

### CODE REVIEW

**Comment by maolson-msft** [comment]
Since RFC4653 doesn't seem like it would map to QUIC as readily as the stuff in the RACK draft about reorder-threshold tuning (it seems to be all about extending the Limited Transmit algorithm), maybe RACK is a better reference?

https://tools.ietf.org/pdf/draft-ietf-tcpm-rack-08.pdf
c.f. "Update RACK reordering window"

**Comment by ianswett** [comment]
I would prefer to reference RACK, but it's not an RFC yet and I'm not sure it will be prior to QUIC shipping.  What's the best way to handle that?

**Comment by maolson-msft** [comment]
The QUIC recovery spec already references the RACK draft in multiple places.

**Comment by ianswett** [comment]
Fair point, I replaced the old reference with the RACK reference.  PTAL.

**Comment by maolson-msft** [comment]
LG!

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -420,8 +420,9 @@ The RECOMMENDED initial value for the packet reordering threshold
 less than 3, to keep in line with TCP {{?RFC5681}}.
 
 Some networks may exhibit higher degrees of reordering, causing a sender to
-detect spurious losses.  Implementers MAY use algorithms developed for TCP, such
-as TCP-NCR {{?RFC4653}}, to improve QUIC's reordering resilience.
+detect spurious losses.  Increasing reordering thresholds after losses
+are spuriously detected have proven useful in TCP, such	as TCP-NCR
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:424]:
```suggestion
are spuriously detected have proven useful in TCP, such as TCP-NCR
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -420,8 +420,9 @@ The RECOMMENDED initial value for the packet reordering threshold
 less than 3, to keep in line with TCP {{?RFC5681}}.
 
 Some networks may exhibit higher degrees of reordering, causing a sender to
-detect spurious losses.  Implementers MAY use algorithms developed for TCP, such
-as TCP-NCR {{?RFC4653}}, to improve QUIC's reordering resilience.
+detect spurious losses.  Increasing reordering thresholds after losses
+are spuriously detected have proven useful in TCP, such as TCP-NCR
+{{?RFC4653}}, and are expected to be equally useful in QUIC.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:425]: The "such as" is injected at an odd point here, which seems to break up the flow.

```suggestion
detect spurious losses.  Increasing reordering thresholds after losses, as
recommended in TCP-NCR {{?RFC4653}}, have proven to be useful in TCP and are
expected to be equally useful in QUIC.
```

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -420,10 +420,9 @@ The RECOMMENDED initial value for the packet reordering threshold
 less than 3, to keep in line with TCP {{?RFC5681}}.
 
 Some networks may exhibit higher degrees of reordering, causing a sender to
-detect spurious losses.  Implementers SHOULD increase QUIC's reordering
-thresholds in packets or time after losses are spuriously detected,
-and can build on mechanisms described in {{?RACK}}.
-
+detect spurious losses.  Increasing reordering thresholds after losses
+are spuriously detected have proven useful in TCP, such	as TCP-NCR
+{{?RFC4653}}, and are expected to be equally useful in QUIC.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
detect spurious losses. Increasing reordering thresholds after spurious
detection of losses can make the transport more robust to reordering,
as TCP-NCR {{?RFC4653}} does for TCP. Such mechanisms are
expected to make QUIC more robust to reordering as well.
```

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md]: never mind this -- I like @martinthomson's suggestion better.

**Comment by maolson-msft** [comment]
I don't see any conversation in this PR about it, but I see a commit titled simply "non-normative" reverting the reference from RACK back to NCR. It's not a big deal, but I'm just curious about the reasoning.

**Comment by ianswett** [comment]
There's conversation on the issue(https://github.com/quicwg/base-drafts/issues/3571#issuecomment-611350198) and Jana was concerned about adding another reference to RACK as well as recommending something that we don't have a detailed algorithm(and pseudocode) for.  In general, anything we have a SHOULD for has more text(and typically pseudocode) than this, so I accepted that point and adjusted the PR.  For example, the pseudocode doesn't detail how to detect spurious losses, since it's expected that could be fairly complex and implementation dependent.

**Comment by maolson-msft** [comment]
I see, thanks for pointing me to that. But Jana's ask seems to me to be all about SHOULD versus MAY, and orthogonal to referencing NCR vs RACK; if it was a SHOULD, then you'd still have the same incompletely-specified-algorithm problem with NCR. It seems to me that now that SHOULD is removed, Jana would agree that RACK is a better reference since it maps more cleanly to QUIC.

**Comment by maolson-msft** [comment]
Actually, I'm not so confident that RACK maps all that cleanly either, and Jana made the point that RACK's ReoWnd adjustment algorithm hasn't been closely examined, so I'm going to bow out here- I'm happy with either NCR or RACK.

**Pull request merged**